src/gentropy/datasource/biosample_ontologies/utils.py

```diff
@@ -0,0 +1,197 @@
+    df_graphs = df.select(explode_outer("graphs").alias("graph"))
+
+    # Exploding the 'nodes' array within each graph
+    df_nodes = df_graphs.select(
+        col("graph.id").alias("graph_id"),
+        explode_outer("graph.nodes").alias("node"))
+
+    # Exploding the 'edges' array within each graph for relationship data
+    df_edges = df_graphs.select(
+        col("graph.id").alias("graph_id"),
+        explode_outer("graph.edges").alias("edge")
+    ).select(
+        col("edge.sub").alias("subject"),
+        col("edge.pred").alias("predicate"),
+        col("edge.obj").alias("object")
+    )
+    df_edges = df_edges.withColumn("subject", regexp_replace(col("subject"), "http://purl.obolibrary.org/obo/", ""))
+    df_edges = df_edges.withColumn("object", regexp_replace(col("object"), "http://purl.obolibrary.org/obo/", ""))
+
+    # Extract the relevant information from the nodes
+    transformed_df = df_nodes.select(
+    regexp_replace(col("node.id"), "http://purl.obolibrary.org/obo/", "").alias("biosampleId"),
+    col("node.lbl").alias("biosampleName"),
+    col("node.meta.definition.val").alias("description"),
+    collect_set(col("node.meta.xrefs.val")).over(Window.partitionBy("node.id")).getItem(0).alias("xrefs"),
+    # col("node.meta.deprecated").alias("deprecated"),
+    collect_set(col("node.meta.synonyms.val")).over(Window.partitionBy("node.id")).getItem(0).alias("synonyms"))
+
+
+    # Extract the relationships from the edges
+    # Prepare relationship-specific DataFrames
+    df_parents = df_edges.filter(col("predicate") == "is_a").select("subject", "object").withColumnRenamed("object", "parent")
+    df_children = df_edges.filter(col("predicate") == "is_a").select("object", "subject").withColumnRenamed("subject", "child")
+
+    # Aggregate relationships back to nodes
+    df_parents_grouped = df_parents.groupBy("subject").agg(array_distinct(collect_list("parent")).alias("parents"))
+    df_children_grouped = df_children.groupBy("object").agg(array_distinct(collect_list("child")).alias("children"))
+
+    # Get all ancestors
+    df_with_ancestors = json_graph_traversal(df_parents_grouped, "subject", "parents", "ancestors")
+    # Get all descendants
+    df_with_descendants = json_graph_traversal(df_children_grouped, "object", "children", "descendants")
+
+    # Join the ancestor and descendant DataFrames
+    df_with_relationships = df_with_ancestors.join(df_with_descendants, df_with_ancestors.subject == df_with_descendants.object, "full_outer").withColumn("biosampleId", coalesce(df_with_ancestors.subject, df_with_descendants.object)).drop("subject", "object")
+
+    # Join the original DataFrame with the relationship DataFrame
+    final_df = transformed_df.join(df_with_relationships, ["biosampleId"], "left")
+
+    return BiosampleIndex(
+        _df=final_df,
+        _schema=BiosampleIndex.get_schema()
+        )
+
+def merge_biosample_indices(
+    biosample_indices : list[BiosampleIndex]
+    ) -> BiosampleIndex:
```

Comment:
I wouldn't add this function here. The merge of biosample indices is kind of an inherent ability of biosample indices as well. Eg.

```python
a: BiosampleIndex
b: BiosampleIndex
merged: BiosampleIndex = a.merge_indices(b)
```
Assuming the operation is commutative, and `b.merge(a)` yields the same index. 

So my advise is to move this logic as a instance method of the `BiosampleIndex` dataset.

Reply:
(also this would imply one fewer function to import, as the merge function travels wherever the biosample object goes)